internal/route/request_executor.go

```diff
@@ -157,64 +168,76 @@ func (r *RequestExecutor) retrieveOrCacheRequest(httpReq *http.Request, requestB
 		case *HandledError:
 			// The cache uses request coalescing, an error may be returned by another goroutine.
 			// Construct a responsebody and a fake response.
-			if resp == nil && respBody == nil {
-				resp = &http.Response{
+			if httpResp == nil && jsonRPCRespBody == nil {
+				httpResp = &HttpResponse{
 					StatusCode: http.StatusOK,
-					Body:       io.NopCloser(new(bytes.Buffer)),
 				}
 				rb := *err.rb
 				rb.ID = *requestBody.ID
 				rb.JSONRPC = requestBody.JSONRPCVersion
-				respBody = &rb
+				jsonRPCRespBody = &rb
 			}
 		default:
 			return nil, nil, cached, err
 		}
 	}
 
-	if resp == nil && respBody == nil {
+	if httpResp == nil && jsonRPCRespBody == nil {
 		if val == nil {
 			return nil, nil, cached, fmt.Errorf("unexpected empty response from cache")
 		}
 
 		r.logger.Debug("cache hit", zap.Any("request", requestBody), zap.Any("value", val))
 
 		// Fill in id and jsonrpc in the respBody to match the request.
-		resp = &http.Response{
+		httpResp = &HttpResponse{
 			StatusCode: http.StatusOK,
-			Body:       io.NopCloser(new(bytes.Buffer)),
 		}
-		respBody = &jsonrpc.SingleResponseBody{
+		jsonRPCRespBody = &jsonrpc.SingleResponseBody{
 			ID:      *requestBody.ID,
 			JSONRPC: requestBody.JSONRPCVersion,
 			Result:  val,
 		}
 	}
 
-	return respBody, resp, cached, nil
+	return jsonRPCRespBody, httpResp, cached, nil
 }
 
-func (r *RequestExecutor) getResponseBody(httpReq *http.Request, requestBody jsonrpc.RequestBody, configToRoute *config.UpstreamConfig) (jsonrpc.ResponseBody, *http.Response, error) {
+func (r *RequestExecutor) getResponseBody(httpReq *http.Request, requestBody jsonrpc.RequestBody, configToRoute *config.UpstreamConfig) (jsonrpc.ResponseBody, *HttpResponse, error) {
 	resp, err := r.httpClient.Do(httpReq)
 
 	if err != nil {
 		r.logger.Error("Error encountered when executing request.", zap.Any("request", requestBody),
-			zap.String("upstreamID", configToRoute.ID), zap.String("response", fmt.Sprintf("%v", resp)), zap.Error(err))
-		return nil, nil, &OriginError{err}
+			zap.String("upstreamID", configToRoute.ID), zap.Error(err))
+
+		return nil, nil, &OriginError{err, "", 0}
 	}
+
+	// Body can only be read once. Read it out and put it back in the response.
+	respBodyString := util.ReadAndCopyBackResponseBody(resp)
+	httpResp := &HttpResponse{resp.StatusCode}
+
+	if resp.StatusCode >= http.StatusBadRequest {
```

Comment:
We weren't doing a good job of handling HTTP responses. 

HTTP responses with non 2xx//3xx codes should count as errors IMO.

Reply:
We were categorizing pretty much any HTTP error as a 500.